pkg/operator/csidriveroperator/deploymentcontroller.go

```diff
@@ -0,0 +1,207 @@
+
+	c := &CSIDriverOperatorDeploymentController{
+		name:              name,
+		operatorClient:    clients.OperatorClient,
+		csiOperatorConfig: csiOperatorConfig,
+		kubeClient:        clients.KubeClient,
+		versionGetter:     versionGetter,
+		targetVersion:     targetVersion,
+		eventRecorder:     eventRecorder,
+		factory:           f,
+	}
+	return c
+}
+
+func (c *CSIDriverOperatorDeploymentController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("CSIDriverOperatorDeploymentController sync started")
+	defer klog.V(4).Infof("CSIDriverOperatorDeploymentController sync finished")
+
+	opSpec, opStatus, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	required := c.getRequiredDeployment(opSpec)
+	lastGeneration := resourcemerge.ExpectedDeploymentGeneration(required, opStatus.Generations)
+	deployment, _, err := resourceapply.ApplyDeployment(c.kubeClient.AppsV1(), c.eventRecorder, required, lastGeneration)
+	if err != nil {
+		klog.V(2).Infof("ApplyDeployment failed: %s", err)
+		// This will set Degraded condition
+		return err
+	}
+	klog.V(4).Infof("ApplyDeployment succeeded")
+
+	// Available: at least one replica is running
+	deploymentAvailable := operatorapi.OperatorCondition{
+		Type: c.name + deploymentControllerName + operatorapi.OperatorStatusTypeAvailable,
+	}
+
+	if deployment.Status.AvailableReplicas > 0 {
+		deploymentAvailable.Status = operatorapi.ConditionTrue
+	} else {
+		deploymentAvailable.Status = operatorapi.ConditionFalse
+		deploymentAvailable.Reason = "WaitDeployment"
+		deploymentAvailable.Message = "Waiting for a Deployment pod to start"
+	}
+
+	// Not progressing: all replicas are at the latest version && Deployment generation matches
+	deploymentProgressing := operatorapi.OperatorCondition{
+		Type: c.name + deploymentControllerName + operatorapi.OperatorStatusTypeProgressing,
+	}
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		deploymentProgressing.Status = operatorapi.ConditionTrue
+		deploymentProgressing.Reason = "NewGeneration"
+		msg := fmt.Sprintf("desired generation %d, current generation %d", deployment.Generation, deployment.Status.ObservedGeneration)
+		deploymentProgressing.Message = msg
+	} else {
+		if deployment.Spec.Replicas != nil {
+			if deployment.Status.UpdatedReplicas == *deployment.Spec.Replicas {
+				deploymentProgressing.Status = operatorapi.ConditionFalse
+				// All replicas were updated, set the version
+				c.versionGetter.SetVersion(c.name+versionName, c.targetVersion)
+			} else {
+				msg := fmt.Sprintf("%d out of %d pods running", deployment.Status.UpdatedReplicas, *deployment.Spec.Replicas)
+				deploymentProgressing.Status = operatorapi.ConditionTrue
+				deploymentProgressing.Reason = "WaitDeployment"
+				deploymentProgressing.Message = msg
+			}
+		}
+	}
```

Comment:
What about Degraded? Did you mean to use `WithSyncDegradedOnError` in line 186?

Reply:
Fixed (also in crcontroller.go)